chirpstack/src/maccommand/mod.rs

```diff
@@ -126,6 +133,156 @@ pub async fn handle_uplink<'a>(
+
+    for cid in cids {
+        let unack_count = get_maccommand_unack_count(cid, ds);
+        if unack_count <= max_unack_count {
+            let last_attempt = unack_count == max_unack_count;
+            if check_nonresponsive_mac_command(&dev_eui, cid, ds).await {
+                warn!(dev_eui = %dev_eui, cid = %cid, unack_count = %unack_count, max_unack_count = %max_unack_count, "Device did not respond to mac-command");
+
+                if last_attempt {
+                    let pl = integration_pb::LogEvent {
+                        time: Some(ts.into()),
+                        device_info: Some(di.clone()),
+                        level: integration_pb::LogLevel::Warning.into(),
+                        code: integration_pb::LogCode::MacCommand.into(),
+                        description: format!("Device not responding to MAC command"),
+                        context: [(
+                            "deduplication_id".to_string(),
+                            uplink_frame_set.uplink_set_id.to_string(),
+                        ),
+                        (
+                            "cid".to_string(),
+                            cid.to_string(),
+                        )]
+                        .iter()
+                        .cloned()
+                        .collect(),
+                    };
+                    integration::log_event(app.id, &dev.variables, &pl).await;
+                }
+            }
+        }
+    }
+}
+
+fn get_maccommand_unack_count(cid: lrwn::CID, ds: &mut internal::DeviceSession) -> u32 {
+    return ds.mac_command_unack_count.get(&(cid.to_u8() as u32))
+              .cloned()
+              .unwrap_or_default();
+}
+
+async fn check_nonresponsive_mac_command(dev_eui: &EUI64, cid: lrwn::CID, ds: &mut internal::DeviceSession) -> bool {
+    // Get pending mac-command block, this could return None.
+    let pending = match mac_command::get_pending(&dev_eui, cid).await {
+        Ok(v) => v,
+        Err(e) => {
+            error!(dev_eui = %dev_eui, cid = %cid, error = %e, "Get pending mac-command block error");
+            return false;
+        }
+    };
+
+    // The device failed to respond to the command. Increment the unacked counter.
+    if pending.is_some() {
+        // increase the unacked counter
+        let count = ds
+            .mac_command_unack_count
+            .entry(cid.to_u8() as u32)
+            .or_insert(0);
+        *count += 1;
+
+        if let Err(e) = mac_command::delete_pending(&dev_eui, cid).await {
+            error!(dev_eui = %dev_eui, cid = %cid, error = %e, "Delete pending mac-command error");
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+// Clears all pending MAC commands for the specified device.
+pub async fn clear_pending_mac_commands(
+    dev_eui: &EUI64,
+) {
+    // A list of MAC commands that the device is expected to respond to, if pending.
+    let cids: Vec<lrwn::CID> = Vec::from([
```

Comment:
This list is going to get out-of-sync as soon as new mac-commands will get introduced. As well the number of Redis queries will increment with each mac-command that will get added to the list. This is going to affect the performance...